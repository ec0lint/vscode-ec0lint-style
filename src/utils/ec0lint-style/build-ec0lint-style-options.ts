import path from 'path';
import pathIsInside from 'path-is-inside';
import { URI } from 'vscode-uri';
import { findPackageRoot } from '../packages';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type { RunnerOptions } from './types';

/**
 * Given a document URI, base options, and extension options, builds a ec0lint-style
 * options object. Runner options supersede base options.
 */
export async function buildEc0lintStyleOptions(
	uri: string,
	workspaceFolder?: string,
	baseOptions: Partial<Ec0lintStyle.LinterOptions> = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportDescriptionlessDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	}: RunnerOptions = {},
): Promise<Partial<Ec0lintStyle.LinterOptions>> {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportDescriptionlessDisables:
			reportDescriptionlessDisables ?? baseOptions.reportDescriptionlessDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.ec0lint-styleignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.ec0lint-styleignore',
			);
		}
	}

	return options;
}
