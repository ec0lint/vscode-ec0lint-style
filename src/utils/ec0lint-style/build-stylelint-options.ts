import path from 'path';
import pathIsInside from 'path-is-inside';
import { URI } from 'vscode-uri';
import { findPackageRoot } from '../packages';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import type { RunnerOptions } from './types';

/**
 * Given a document URI, base options, and extension options, builds a Stylelint
 * options object. Runner options supersede base options.
 */
export async function buildStylelintOptions(
	uri: string,
	workspaceFolder?: string,
	baseOptions: Partial<Ec0lintStyle.LinterOptions> = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportDescriptionlessDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	}: RunnerOptions = {},
): Promise<Partial<Ec0lintStyle.LinterOptions>> {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportDescriptionlessDisables:
			reportDescriptionlessDisables ?? baseOptions.reportDescriptionlessDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.stylelintignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.stylelintignore',
			);
		}
	}

	return options;
}
