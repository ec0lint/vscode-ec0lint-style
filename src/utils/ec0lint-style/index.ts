export * from './build-stylelint-options';
export * from './disable-metadata-lookup-table';
export * from './formatting-options-to-rules';
export * from './get-disable-diagnostic-rule';
export * from './is-disable-report-rule';
export * from './process-linter-result';
export * from './stylelint-runner';
export * from './types';
export * from './warning-to-diagnostic';
