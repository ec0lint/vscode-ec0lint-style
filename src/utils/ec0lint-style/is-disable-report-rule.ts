import { DisableReportRuleNames } from './types';

/**
 * Returns whether or not the given rule is a disable report rule.
 * @param rule The rule to check.
 */
export function isDisableReportRule(rule: unknown): rule is DisableReportRuleNames {
	switch (rule as DisableReportRuleNames) {
		case DisableReportRuleNames.Descriptionless:
		case DisableReportRuleNames.Illegal:
		case DisableReportRuleNames.InvalidScope:
		case DisableReportRuleNames.Needless:
			return true;

		default:
			return false;
	}
}
