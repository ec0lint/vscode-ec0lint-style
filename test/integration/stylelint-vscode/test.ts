import { join, resolve } from 'path';
import { pathToFileURL } from 'url';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { Ec0lintRunner as Ec0lintStyleRunner } from '../../../src/utils/ec0lint-style';

const createDocument = (uri: string | null, languageId: string, contents: string): TextDocument =>
	TextDocument.create(
		uri ? pathToFileURL(resolve(__dirname, '..', uri)).toString() : 'Untitled:Untitled',
		languageId,
		1,
		contents,
	);

describe('StylelintRunner', () => {
	test('should be resolved with diagnostics when it lints CSS successfully', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument(null, 'css', '  a[id="id"]{}'), {
			config: {
				rules: {
					'string-quotes': ['single', { severity: 'warning' }],
					indentation: ['tab'],
				},
			},
		});

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved with an empty array when no errors and warnings are reported', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument(null, 'scss', ''), {
			config: {
				customSyntax: 'postcss-scss',
				rules: { indentation: [2] },
			},
		});

		expect(result.diagnostics).toEqual([]);
	});

	test('should be resolved with one diagnostic when the CSS is broken', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		// TODO: Restore once postcss-markdown is PostCSS 8 compatible
		// 		const result = await runner.lintDocument(
		// 			createDocument(
		// 				'markdown.md',
		// 				'markdown',
		// 				`# Title

		// # Code block

		// \`\`\`css
		//           a{
		// \`\`\`
		// `,
		// 			),
		// 			{
		// 				config: {
		// 					customSyntax: 'postcss-markdown',
		// 					rules: {
		// 						indentation: ['tab'],
		// 					},
		// 				},
		// 			},
		// 		);
		const result = await runner.lintDocument(
			createDocument('scss.scss', 'scss', '          a{\n'),
			{
				config: {
					customSyntax: 'postcss-scss',
					rules: {
						indentation: ['tab'],
					},
				},
			},
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved even if no configs are defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		// TODO: Restore once postcss-html is PostCSS 8 compatible
		// const result = await runner.lintDocument(createDocument(null, 'plaintext', '<style>a{}</style>'), {
		// 	customSyntax: 'postcss-html',
		// });
		const result = await runner.lintDocument(createDocument(null, 'plaintext', 'a{}'), {
			customSyntax: 'postcss-scss',
		});

		expect(result.diagnostics).toEqual([]);
	});

	test('should support `.stylelintignore`.', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('listed-in-stylelintignore.css', 'css', '}'),
			{
				ignorePath: require.resolve('./.stylelintignore'),
			},
		);

		expect(result.diagnostics).toEqual([]);
	});

	test('should support CSS-in-JS with customSyntax', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				null,
				'javascript',
				`import glamorous from 'glamorous';
const styled = require("styled-components");
const A = glamorous.a({font: 'bold'});
const B = styled.b\`
font: normal
\`;`,
			),
			{
				config: {
					customSyntax: '@stylelint/postcss-css-in-js',
					rules: { 'font-weight-notation': ['numeric'] },
				},
			},
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should set `codeFilename` option from a TextDocument', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'should-be-ignored.xml',
				'xml',
				`<style>
a { color: #000 }
</style>`,
			),
			{
				config: {
					rules: {
						'color-hex-length': 'long',
					},
					ignoreFiles: '**/*-ignored.xml',
				},
			},
		);

		expect(result.diagnostics).toEqual([]);
	});

	test('should support `processors` option', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('processors.tsx', 'typescriptreact', 'styled.p`"`'),
			{
				config: {
					processors: ['stylelint-processor-styled-components'],
					rules: {},
				},
			},
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should check CSS syntax even if no configuration is provided', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('unclosed.css', 'css', 'a{color:rgba(}'),
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should check CSS syntax even if no rule is provided', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		// TODO: Restore once postcss-html is PostCSS 8 compatible
		// const result = await runner.lintDocument(createDocument('at.xsl', 'xsl', '<style>@</style>'), {
		// 	customSyntax: 'postcss-html',
		// });
		const result = await runner.lintDocument(createDocument('at.scss', 'scss', '@'), {
			customSyntax: 'postcss-scss',
		});

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved even if no rules are defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument('no-rules.css', 'css', 'a{}'), {
			config: {},
		});

		expect(result.diagnostics).toEqual([]);
	});

	test('should reject with a reason when it takes incorrect options', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const promise = runner.lintDocument(
			createDocument('invalid-options.css', 'css', '  a[id="id"]{}'),
			{
				config: {
					rules: {
						'string-quotes': 'single',
						'color-hex-case': 'foo',
						'at-rule-empty-line-before': ['always', { bar: true }],
					},
				},
			},
		);

		await expect(promise).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should be resolved with diagnostics when the rules include unknown rules', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument('unknown-rule.css', 'css', 'b{}'), {
			config: {
				rules: {
					'this-rule-does-not-exist': 1,
					'this-rule-also-does-not-exist': 1,
				},
			},
		});

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved with diagnostic plugin rule URL', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('unknown-rule.scss', 'scss', '@unknown (max-width: 960px) {}'),
			{
				config: {
					plugins: ['stylelint-scss'],
					rules: {
						'scss/at-rule-no-unknown': true,
					},
				},
			},
		);

		expect(result.diagnostics).toMatchSnapshot();
	});
});

describe('StylelintRunner with a configuration file', () => {
	test('should adhere to configuration file settings', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				join(__dirname, 'has-config-file.tsx'),
				'typescriptreact',
				`
const what: string = "is this";
<a css={{
  width: "0px",
  what
}} />;
`,
			),
			{ configFile: join(__dirname, 'no-unknown.config.js') },
		);

		expect(result.diagnostics).toMatchSnapshot();
	});
});

describe('StylelintRunner with auto-fix', () => {
	test('auto-fix should work properly if configs are defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument(null, 'css', 'a\n{\ncolor:red;\n}'), {
			config: { rules: { indentation: [2] } },
			fix: true,
		});

		expect(result.output).toMatchSnapshot();
	});

	test('auto-fix should only work properly for syntax errors if no rules are defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument('no-rules.css', 'css', 'a {'), {
			config: {},
			fix: true,
		});

		expect(result.output).toMatchInlineSnapshot(`"a {}"`);
	});

	test('JS file auto-fix should not change the content if no rules are defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument('no-rules.js', 'javascript', '"a"'), {
			customSyntax: '@stylelint/postcss-css-in-js',
			config: {},
			fix: true,
		});

		expect(result.output).toMatchInlineSnapshot(`"\\"a\\""`);
	});

	test('auto-fix should ignore if the file matches the ignoreFiles', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('should-be-ignored.js', 'javascript', '"a"'),
			{
				customSyntax: '@stylelint/postcss-css-in-js',
				config: {
					rules: {},
					ignoreFiles: '**/*-ignored.js',
				},
				fix: true,
			},
		);

		expect(result.output).toBeUndefined();
	});

	test('auto-fix should work if there is syntax errors in css', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
.a {
	width: 100%
	height: 100%;
}
`,
			),
			{
				config: { rules: {} },
				fix: true,
			},
		);

		expect(result.output).toMatchSnapshot();
	});

	test('auto-fix should ignore if there is syntax errors in scss', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.scss',
				'scss',
				`
.a {
	width: 100%
	height: 100%;
}
`,
			),
			{
				customSyntax: 'postcss-scss',
				config: { rules: {} },
				fix: true,
			},
		);

		expect(result.output).toBeUndefined();
	});

	test('auto-fix should work if there are errors that cannot be auto-fixed', async () => {
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
unknown {
    width: 100%;
    height: 100%;
}
`,
			),
			{
				config: {
					rules: {
						indentation: 2,
						'selector-type-no-unknown': true,
					},
				},
				fix: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with customSyntax', () => {
	test('should work properly if customSyntax is defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument('test.css', 'css', 'a\n   color:red'), {
			config: { rules: { indentation: [2] } },
			customSyntax: 'postcss-sass',
		});

		expect(result).toMatchSnapshot();
	});

	test('auto-fix should work properly if customSyntax is defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(createDocument('test.css', 'css', 'a\n   color:red'), {
			config: { rules: { indentation: [2] } },
			customSyntax: 'postcss-sass',
			fix: true,
		});

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with reportDescriptionlessDisables', () => {
	test('should work properly if reportDescriptionlessDisables is true', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
.baz {
    /* stylelint-disable-next-line indentation */
  color: red;
}
/* stylelint-disable indentation */
.baz {
  color: red;
}
/* stylelint-enable indentation */
.baz {
  color: red; /* stylelint-disable-line indentation */
}

.baz {
    /* stylelint-disable-next-line indentation -- with a description */
  color: red;
}
/* stylelint-disable indentation -- with a description */
.baz {
  color: red;
}
/* stylelint-enable indentation */
.baz {
  color: red; /* stylelint-disable-line indentation -- with a description */
}
`,
			),
			{
				config: { rules: { indentation: [4] } },
				reportDescriptionlessDisables: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with reportNeedlessDisables', () => {
	test('should work properly if reportNeedlessDisables is true', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
.foo {
  /* stylelint-disable-next-line indentation */
    color: red;
}

/* stylelint-disable indentation */
.bar {
    color: red;
}
/* stylelint-enable indentation */

.baz {
    color: red; /* stylelint-disable-line indentation */
}

/* stylelint-disable indentation */
.bar {
    color: red;
}
`,
			),
			{
				config: { rules: { indentation: [4] } },
				reportNeedlessDisables: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with reportInvalidScopeDisables', () => {
	test('should work properly if reportInvalidScopeDisables is true', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
/* stylelint-disable-next-line foo */

/* stylelint-disable-line foo */

/* stylelint-disable foo */
/* stylelint-enable foo */

/* stylelint-disable-next-line indentation */

/* stylelint-disable-line indentation */

/* stylelint-disable indentation */
/* stylelint-enable indentation */
`,
			),
			{
				config: { rules: { indentation: [4] } },
				reportInvalidScopeDisables: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with stylelintPath', () => {
	test('should work properly if stylelintPath is defined', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('test.css', 'css', 'a{\n   color:red}'),
			{
				config: { rules: { indentation: [2] } },
			},
			{
				stylelintPath: resolve(__dirname, '../../../node_modules/ec0lint-style'),
			},
		);

		expect(result).toMatchSnapshot();
	});

	test('should work properly if custom path is defined in stylelintPath', async () => {
		expect.assertions(1);
		const runner = new Ec0lintStyleRunner();
		const result = await runner.lintDocument(
			createDocument('test.css', 'css', 'a{\n   color:red}'),
			{
				config: { rules: { indentation: [2] } },
			},
			{
				stylelintPath: require.resolve('./fake-stylelint'),
			},
		);

		expect(result).toMatchSnapshot();
	});
});
