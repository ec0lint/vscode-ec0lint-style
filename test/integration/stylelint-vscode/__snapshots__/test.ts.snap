// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StylelintRunner should be resolved with diagnostic plugin rule URL 1`] = `
Array [
  Object {
    "code": "scss/at-rule-no-unknown",
    "codeDescription": Object {
      "href": "https://github.com/stylelint-scss/stylelint-scss/blob/master/src/rules/at-rule-no-unknown",
    },
    "message": "Unexpected unknown at-rule \\"@unknown\\" (scss/at-rule-no-unknown)",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should be resolved with diagnostics when it lints CSS successfully 1`] = `
Array [
  Object {
    "code": "string-quotes",
    "codeDescription": Object {
      "href": "https://stylelint.io/user-guide/rules/list/string-quotes",
    },
    "message": "Expected single quotes (string-quotes)",
    "range": Object {
      "end": Object {
        "character": 8,
        "line": 0,
      },
      "start": Object {
        "character": 7,
        "line": 0,
      },
    },
    "severity": 2,
    "source": "ec0lint-style",
  },
  Object {
    "code": "indentation",
    "codeDescription": Object {
      "href": "https://stylelint.io/user-guide/rules/list/indentation",
    },
    "message": "Expected indentation of 0 tabs (indentation)",
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 0,
      },
      "start": Object {
        "character": 2,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should be resolved with diagnostics when the rules include unknown rules 1`] = `
Array [
  Object {
    "code": "this-rule-does-not-exist",
    "message": "Unknown rule this-rule-does-not-exist.",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
  Object {
    "code": "this-rule-also-does-not-exist",
    "message": "Unknown rule this-rule-also-does-not-exist.",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should be resolved with one diagnostic when the CSS is broken 1`] = `
Array [
  Object {
    "code": "CssSyntaxError",
    "message": "Unclosed block (CssSyntaxError)",
    "range": Object {
      "end": Object {
        "character": 11,
        "line": 0,
      },
      "start": Object {
        "character": 10,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should check CSS syntax even if no configuration is provided 1`] = `
Array [
  Object {
    "code": "CssSyntaxError",
    "message": "Unclosed bracket (CssSyntaxError)",
    "range": Object {
      "end": Object {
        "character": 13,
        "line": 0,
      },
      "start": Object {
        "character": 12,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should check CSS syntax even if no rule is provided 1`] = `
Array [
  Object {
    "code": "CssSyntaxError",
    "message": "At-rule without name (CssSyntaxError)",
    "range": Object {
      "end": Object {
        "character": 1,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should reject with a reason when it takes incorrect options 1`] = `
"Invalid option name \\"bar\\" for rule \\"at-rule-empty-line-before\\"
Invalid option value \\"foo\\" for rule \\"color-hex-case\\""
`;

exports[`StylelintRunner should support \`processors\` option 1`] = `
Array [
  Object {
    "code": "CssSyntaxError",
    "message": "Unclosed string (CssSyntaxError)",
    "range": Object {
      "end": Object {
        "character": 10,
        "line": 0,
      },
      "start": Object {
        "character": 9,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner should support CSS-in-JS with customSyntax 1`] = `
Array [
  Object {
    "code": "font-weight-notation",
    "codeDescription": Object {
      "href": "https://stylelint.io/user-guide/rules/list/font-weight-notation",
    },
    "message": "Expected numeric font-weight notation (font-weight-notation)",
    "range": Object {
      "end": Object {
        "character": 30,
        "line": 2,
      },
      "start": Object {
        "character": 29,
        "line": 2,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
  Object {
    "code": "font-weight-notation",
    "codeDescription": Object {
      "href": "https://stylelint.io/user-guide/rules/list/font-weight-notation",
    },
    "message": "Expected numeric font-weight notation (font-weight-notation)",
    "range": Object {
      "end": Object {
        "character": 7,
        "line": 4,
      },
      "start": Object {
        "character": 6,
        "line": 4,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner with a configuration file should adhere to configuration file settings 1`] = `
Array [
  Object {
    "code": "length-zero-no-unit",
    "codeDescription": Object {
      "href": "https://stylelint.io/user-guide/rules/list/length-zero-no-unit",
    },
    "message": "Unexpected unit (length-zero-no-unit)",
    "range": Object {
      "end": Object {
        "character": 11,
        "line": 3,
      },
      "start": Object {
        "character": 10,
        "line": 3,
      },
    },
    "severity": 1,
    "source": "ec0lint-style",
  },
]
`;

exports[`StylelintRunner with auto-fix auto-fix should work if there are errors that cannot be auto-fixed 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "selector-type-no-unknown",
      "codeDescription": Object {
        "href": "https://stylelint.io/user-guide/rules/list/selector-type-no-unknown",
      },
      "message": "Unexpected unknown type selector \\"unknown\\" (selector-type-no-unknown)",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 1,
        },
        "start": Object {
          "character": 0,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
  "output": "
unknown {
  width: 100%;
  height: 100%;
}
",
}
`;

exports[`StylelintRunner with auto-fix auto-fix should work if there is syntax errors in css 1`] = `
"
.a {
	width: 100%;
	height: 100%
}
"
`;

exports[`StylelintRunner with auto-fix auto-fix should work properly if configs are defined 1`] = `
"a
{
  color:red;
}"
`;

exports[`StylelintRunner with customSyntax auto-fix should work properly if customSyntax is defined 1`] = `
Object {
  "diagnostics": Array [],
  "output": "a
  color:red",
}
`;

exports[`StylelintRunner with customSyntax should work properly if customSyntax is defined 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "indentation",
      "codeDescription": Object {
        "href": "https://stylelint.io/user-guide/rules/list/indentation",
      },
      "message": "Expected indentation of 2 spaces (indentation)",
      "range": Object {
        "end": Object {
          "character": 4,
          "line": 1,
        },
        "start": Object {
          "character": 3,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`StylelintRunner with reportDescriptionlessDisables should work properly if reportDescriptionlessDisables is true 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "--report-descriptionless-disables",
      "message": "Disable for \\"indentation\\" is missing a description",
      "range": Object {
        "end": Object {
          "character": 5,
          "line": 2,
        },
        "start": Object {
          "character": 4,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-descriptionless-disables",
      "message": "Disable for \\"indentation\\" is missing a description",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 5,
        },
        "start": Object {
          "character": 0,
          "line": 5,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-descriptionless-disables",
      "message": "Disable for \\"indentation\\" is missing a description",
      "range": Object {
        "end": Object {
          "character": 15,
          "line": 11,
        },
        "start": Object {
          "character": 14,
          "line": 11,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`StylelintRunner with reportInvalidScopeDisables should work properly if reportInvalidScopeDisables is true 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "--report-invalid-scope-disables",
      "message": "Rule \\"foo\\" isn't enabled",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 1,
        },
        "start": Object {
          "character": 0,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-invalid-scope-disables",
      "message": "Rule \\"foo\\" isn't enabled",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 3,
        },
        "start": Object {
          "character": 0,
          "line": 3,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-invalid-scope-disables",
      "message": "Rule \\"foo\\" isn't enabled",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 5,
        },
        "start": Object {
          "character": 0,
          "line": 5,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`StylelintRunner with reportNeedlessDisables should work properly if reportNeedlessDisables is true 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "indentation",
      "codeDescription": Object {
        "href": "https://stylelint.io/user-guide/rules/list/indentation",
      },
      "message": "Expected indentation of 4 spaces (indentation)",
      "range": Object {
        "end": Object {
          "character": 3,
          "line": 2,
        },
        "start": Object {
          "character": 2,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-needless-disables",
      "message": "Needless disable for \\"indentation\\"",
      "range": Object {
        "end": Object {
          "character": 3,
          "line": 2,
        },
        "start": Object {
          "character": 2,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-needless-disables",
      "message": "Needless disable for \\"indentation\\"",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 6,
        },
        "start": Object {
          "character": 0,
          "line": 6,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-needless-disables",
      "message": "Needless disable for \\"indentation\\"",
      "range": Object {
        "end": Object {
          "character": 17,
          "line": 13,
        },
        "start": Object {
          "character": 16,
          "line": 13,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "--report-needless-disables",
      "message": "Needless disable for \\"indentation\\"",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 16,
        },
        "start": Object {
          "character": 0,
          "line": 16,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`StylelintRunner with stylelintPath should work properly if custom path is defined in stylelintPath 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "fake",
      "message": "Fake result",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`StylelintRunner with stylelintPath should work properly if stylelintPath is defined 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "indentation",
      "codeDescription": Object {
        "href": "https://stylelint.io/user-guide/rules/list/indentation",
      },
      "message": "Expected indentation of 2 spaces (indentation)",
      "range": Object {
        "end": Object {
          "character": 4,
          "line": 1,
        },
        "start": Object {
          "character": 3,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;
